### PR TITLE
Removes `type_name` violation for SwiftUI template code

### LINKED ISSUE #2791: `type_name` violation for with SwiftUI template code
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

1. With the new code templates from SwiftUI and XCode previews, to render a View struct with Preview it is necessary to define a `struct YourView_Previews: PreviewProvider`. But to name a struct as such is a `type_name` violation because it uses "_".

2. So it would be nice if we have either this exception of allowing structs that inherit from `Previews` with `_Previews` in the end of the name.

3. This rule should just extend the current `type_name` rule.

4. The rule should be enabled by default.

### PR DESCRIPTION
**Pull request by atfelix**
Closes #2791.

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/Idiomatic/TypeNameRule.swift [hunk 1]
```diff
@@ -68,7 +68,13 @@ public struct TypeNameRule: ASTRule, ConfigurationProviderRule {
             return []
         }
 
+        #if canImport(SwiftUI)
+        let name = name
+            .nameStrippingLeadingUnderscoreIfPrivate(dictionary)
+            .nameStrippingTrailingSwiftUIPreviewProvider(dictionary)
+        #else
         let name = name.nameStrippingLeadingUnderscoreIfPrivate(dictionary)
+        #endif
```
Comment (atfelix) [inline Source/SwiftLintFramework/Rules/Idiomatic/TypeNameRule.swift]: The compiler directives seem to be causing the failures, which are due to `testRulesDocumentationIsUpdated` failing.  Initially there was no compiler directive but there were some significant (~35% slower) performance.  I wasn't sure if I should worry about these performance issues.  Also, anyone on Linux doesn't need this fix.  Any suggestions on how to proceed @marcelofabri, @jpsim?

File: Source/SwiftLintFramework/Rules/Idiomatic/TypeNameRule.swift [hunk 2]
```diff
@@ -92,3 +94,13 @@ public struct TypeNameRule: ASTRule, ConfigurationProviderRule {
         return []
     }
 }
+
+fileprivate extension String {
+    func nameStrippingTrailingSwiftUIPreviewProvider(_ dictionary: [String: SourceKitRepresentable]) -> String {
+        guard dictionary.inheritedTypes.contains("PreviewProvider"),
+            let lastPreviewsIndex = lastIndex(of: "_Previews")
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Idiomatic/TypeNameRule.swift]: I'd make sure that `_Previews` is also the suffix just to be extra careful

File: CHANGELOG.md [hunk 3]
```diff
@@ -28,6 +28,10 @@
 
 #### Enhancements
 
+* Type name rules considers SwiftUI template code.
```
Comment (marcelofabri) [inline CHANGELOG.md]: this is now in the wrong section

File: Source/SwiftLintFramework/Rules/Idiomatic/TypeNameRule.swift [hunk 4]
```diff
@@ -92,3 +94,13 @@ public struct TypeNameRule: ASTRule, ConfigurationProviderRule {
         return []
     }
 }
+
+fileprivate extension String {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/Idiomatic/TypeNameRule.swift]: nitpick: this can be a `private extension`

File: CHANGELOG.md (hunk 3 above)
Comment (atfelix) [reply CHANGELOG.md]: Should it start at Line 13 of this file?

File: Source/SwiftLintFramework/Rules/Idiomatic/TypeNameRule.swift (hunk 4 above)
Comment (atfelix) [reply Source/SwiftLintFramework/Rules/Idiomatic/TypeNameRule.swift]: Thank you.  TIL the above.  That is unexpected behaviour to me.  I thought `private extension` would mean that only `String` would see `nameStrippingTrailingSwiftUIPreviewProvider`

File: Source/SwiftLintFramework/Rules/Idiomatic/TypeNameRule.swift (hunk 2 above)
Comment (atfelix) [reply Source/SwiftLintFramework/Rules/Idiomatic/TypeNameRule.swift]: Okay.  Done.  I'm just waiting for a response to this [comment](https://github.com/realm/SwiftLint/pull/2815/files/1f73dcd0edbfa2a00a69875a505c68583c450078#r317406376) to push the changes

File: CHANGELOG.md (hunk 3 above)
Comment (marcelofabri) [reply CHANGELOG.md]: Yes

**Comment by marcelofabri** [comment]
Thanks!

**Pull request merged**